### PR TITLE
Feat/contract ingestion

### PR DESCRIPTION
**Pull request by geofmureithi**
- Completes Contact ingestion as part of #17

### CODE REVIEW

File: quasar_migrations/src/m20231013_065827_create_contract_table.rs [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+#[derive(DeriveIden)]
+enum Contract {
+    #[sea_orm(iden = "contracts")]
+    Table,
+    Address,
+    Hash,
+    Key,
+    Type,
+    LastModified,
+    // TODO: @geofmureithi Soroban contracts can be introspected.
```
Comment (tomekpiotrowski) [inline quasar_migrations/src/m20231013_065827_create_contract_table.rs:54]: I'd be against keeping TODOs in commited code. We can keep things to do as Github issues and stick to using TODOs only in local code.

What do you think?

File: src/ingestion.rs [hunk 2]
```diff
@@ -22,102 +22,18 @@ pub enum IngestionError {
...
-                    error!("Error while ingesting ledgers: {:?}", error);
-                }
-            }
-            Ok(IngestionNeeded::No) => {}
-            Err(error) => {
-                error!("Error while checking for new ledgers: {}", error);
-            }
-        }
-    }
+    let ledgers = ledgers::ingest(&node_database, &quasar_database, &ingestion, &metrics);
```
Comment (tomekpiotrowski) [inline src/ingestion.rs:31]: There is a potentially significant drawback that this architecture has - it'd be impossible to keep data fully consistent. The old, ledger-based approach had the benefit of it being possible to be wrapped in a single, giant transaction. With separate ingestion processes we're almost guaranteed to have inconsistent data throughout the ingestion process.

I was internally debating whether this is something we should worry about in the PoC, but if there were an easy way of achieving data consistency then we should go for it.

File: quasar_migrations/src/m20231013_065827_create_contract_table.rs (hunk 1 above)
Comment (geofmureithi) [reply quasar_migrations/src/m20231013_065827_create_contract_table.rs:54]: Sure, I will create an issue, this is related to the Discord question, I was waiting to know the next step.

File: src/ingestion.rs (hunk 2 above)
Comment (geofmureithi) [reply src/ingestion.rs:31]: You wanted to wrap the loop in a transaction?

Comment (tomekpiotrowski) [reply src/ingestion.rs:31]: I wanted to wrap `ingest_ledger` in a transaction. This way if there's a new ledger all the new updates for all entities related to that new ledger get applied to the Quasar database in a single transaction. This way the data is always consistent.

Comment (geofmureithi) [reply src/ingestion.rs:31]: Cant we still do it with the current approach?

Comment (geofmureithi) [reply src/ingestion.rs:31]: The changes i did was just move around code.
I am ok if you have suggestions of what code changes make it impossible to do the transaction.

Comment (tomekpiotrowski) [reply src/ingestion.rs:31]: If I understand correctly your code has a separate check for new ledgers for each new entity type. That check is done every 5 seconds for each entity. If I'm not mistaken the following scenario is possible:

1. Transaction ingestion check runs, but finds no new ledgers. No new ledgers are present so nothing is done.
2. 1 second later a new ledger is created.
3. 1 second later the account ingestion checks for a new ledger. It finds a new ledger so it imports accounts.
4. A users runs a GraphQL query to query for accounts. The users tries to join related transactions, but the transaction ingestion hasn't run yet as it'll run in a few seconds. The user receives inconsistent data - accounts without related transactions.

Comment (tomekpiotrowski) [reply src/ingestion.rs:31]: If we have a single entry point to the entire ingestion process, ie. the check for a new ledger is executed in only one place then updates of all entities can be done all at once and encapsulated in a single transaction. So we should have (pseudocode):

```
loop {
  sleep();

  if new_ledger_present {
    start_transaction();
    
    ingest_ledgers();
    ingest_accounts();
    [... other ingestions ...]
    
    end_transaction();
  }
}
```

File: quasar_migrations/src/m20231013_065827_create_contract_table.rs (hunk 1 above)
Comment (geofmureithi) [reply quasar_migrations/src/m20231013_065827_create_contract_table.rs:54]: Documented here: https://github.com/eigerco/quasar/issues/42

File: src/ingestion.rs (hunk 2 above)
Comment (geofmureithi) [reply src/ingestion.rs]: I have made some updates, let me know what you think.

File: quasar_migrations/src/m20231013_065827_create_contract_table.rs (hunk 1 above)
Comment (tomekpiotrowski) [reply quasar_migrations/src/m20231013_065827_create_contract_table.rs:54]: Can you remove this line then?

Comment (geofmureithi) [reply quasar_migrations/src/m20231013_065827_create_contract_table.rs:54]:
```suggestion
```

**Review by tomekpiotrowski — APPROVED** [review]

**Pull request merged**